apps/wallet-api-tools/src/Editor.tsx

```diff
@@ -1,92 +1,159 @@
-    case "in":
-      return "<- [received]";
-    case "out":
-      return "-> [sent]";
-    case "error":
-      return "[error]";
-    case "info":
-      return "[info]";
-    default: {
-      const exhaustiveCheck: never = message.type; // https://www.typescriptlang.org/docs/handbook/2/narrowing.html#exhaustiveness-checking
-      return exhaustiveCheck;
-    }
-  }
-}
+const historyAtom = atomWithStorage<(MessageGrouped | Message)[]>(
+  "history",
+  [],
+);
 
 export function Editor() {
-  const [history, setHistory] = useState<Message[]>([]);
+  const [history, setHistoryAtom] = useAtom(historyAtom);
+  const [initialHistoryLoad, setInitialHistoryLoad] = useState(true);
   const searchParams = useSearchParams();
-  const codeMirrorRef = useRef<ReactCodeMirrorRef | null>(null);
+  const panelRef = useRef<HTMLDivElement>(null);
   const transportRef = useRef<Transport | null>(null);
+  const [value, setValue] = useState("");
+  const [scrollBottom, setScrollBottom] = useState(true);
+
+  //set the scroll to the bottom at first render or when the user is already to bottom
+  useEffect(() => {
+    if (
+      history.length > 0 &&
+      panelRef.current &&
+      (initialHistoryLoad || scrollBottom)
+    ) {
+      const element = panelRef.current;
+      setTimeout(() => {
+        if (panelRef.current !== null) {
+          element.scrollTop = element.scrollHeight;
+        }
+      }, 30);
+      setInitialHistoryLoad(false);
+    }
+  }, [history, initialHistoryLoad]);
+
+  //update scrollBottom when the user scroll
+  useEffect(() => {
+    const element = panelRef.current;
+    if (element) {
+      const updatePosition = () => {
+        const isAtBottom =
+          element.scrollHeight - element.scrollTop === element.clientHeight;
+        setScrollBottom(isAtBottom);
+      };
+      element.addEventListener("scroll", updatePosition);
+      updatePosition();
+
+      return () => element.removeEventListener("scroll", updatePosition);
+    }
+    return undefined;
+  }, []);
 
   const isSimulator = searchParams.get("simulator");
 
   const pushMessage = useCallback(
-    (type: Message["type"], value: Message["value"]) => {
-      setHistory((prev) => [...prev, { date: new Date(), type, value }]);
+    (type: Message["type"], messageValue: Message["value"]) => {
+      const parsedValue: {
+        id: string;
+      } = type === "in" || type === "out" ? JSON.parse(messageValue) : null;
+      if (
+        type === "error" ||
+        type === "info" ||
+        (type === "in" && !parsedValue.id)
+      ) {
+        setHistoryAtom((prev) => [
+          ...prev,
+          { date: new Date(), type, value: messageValue },
+        ]);
+      } else if (type === "out") {
+        setHistoryAtom((prev) => [
+          ...prev,
+          {
+            type: "group",
+            id: parsedValue.id,
+            date: new Date(),
+            messages: {
+              sent: { date: new Date(), type, value: messageValue },
```

Comment:
If we don't want to do the parse again in the history it might be better to store the `parsedValue` instead of `messageValue`, you can type value in a very permissive way

And we should probably do that for every `value` in this fonction